lib/Parse/Lexer.cpp

```diff
@@ -1887,11 +1887,13 @@ void Lexer::lexStringLiteral(unsigned CustomDelimiterLen) {
       } else if ((*CurPtr == '\r' || *CurPtr == '\n') && IsMultilineString) {
         // The only case we reach here is unterminated single line string in the
         // interpolation. For better recovery, go on after emitting an error.
-        diagnose(CurPtr, diag::lex_unterminated_string);
+        diagnose(CurPtr, diag::string_interpolation_unclosed);
         wasErroneous = true;
         continue;
       } else {
-        diagnose(TokStart, diag::lex_unterminated_string);
+        diagnose(CurPtr, diag::string_interpolation_unclosed);
+        diagnose(CurPtr, diag::string_interpolation_unclosed_fix_it)
+            .fixItInsert(Lexer::getSourceLoc(CurPtr), ")");
         return formToken(tok::unknown, TokStart);
```

Comment:
You should probably backtrack to the start of the string delimiter using `QuoteChar` and `CustomDelimiterLen` if you hit one here to correctly compute the insertion location for the closing paren (`"hello \(x` vs. `"hello \(x"`).

Reply:
I’m a bit confused here, but I think I know what you mean

Reply:
Feel free to reach out if you're stuck!

Reply:
Okay so the strategy here would be: look at CustomDelimiterLen, and push CurPtr back by that amount, ie, if CustomDelimeterLen is 2, push the CurPtr back by 2 characters, then diagnose there, correct

Reply:
@AnthonyLatsis Do you think this would work? I don't think it would
```c
        char* tmpDiagnosePtr = CurPtr;
        tmpDiagnosePtr[strlen(tmpDiagnosePtr)-CustomDelimiterLen] = '\0';
        diagnose(tmpDiagnosePtr, diag::string_interpolation_unclosed_fix_it)
            .fixItInsert(Lexer::getSourceLoc(tmpDiagnosePtr), ")");
```

Reply:
Not sure what the purpose of the second line is.

I imagine something like the following, given that `CurPtr` points to a line break or EOF (otherwise, there is likely something else wrong with the interpolation, so falling back to the generic "unterminated string" error is the best we can do right now):
```
char *Loc = CurPtr - 1;
if (*Loc[-1] == QuoteChar) {
  Loc -= CustomDelimiterLen;
}
```

Reply:
Oh I see, will push that and see how if the fix-its given will be accurate